### ISSUE TITLE
Trouble connecting to Bebop on OS X

### ISSUE THREAD

**Issue by simon-o-matic** [Parrot-Developers/ARSDKBuildUtils#5]
Hi Bebob SDK Team,

I'm trying to build a pure OS X client (no iOS, no Android libraries). I've made a minimal program at this point that just starts up the listener threads. I've connected my mac book to the bebop wireless network (and can telnet so I know its running) but my program responds with: 

    ARNETWORKAL_WifiNetwork_Receive:825 - connection lost (too long time without reception)

The basic flow of my code of the sdk calls is:

    ARNETWORKAL_Manager_t* alManager = ARNETWORKAL_Manager_New(&networkAlError);
    ARNETWORKAL_Manager_InitWifiNetwork(alManager, IP_ADDRESS, SENDING_PORT, READING_PORT, NETWORK_RESPONSE_TIMEOUT_SEC); // timeout = 10 seconds
    ARNETWORK_Manager_New(alManager, num_c2d_params, c2d_params, num_d2c_params, d2c_params, PING_DELAY, managerDisconnect, NULL, &networkError);
    ARSAL_Thread_Create(&rxThread, ARNETWORK_Manager_ReceivingThreadRun, netManager);
    ARSAL_Thread_Create(&txThread, ARNETWORK_Manager_SendingThreadRun, netManager);

I sleep for a while and then after 10 seconds I get that error. What am I missing? 

Here is some other info:

    #define SENDING_PORT 	54321
    #define READING_PORT 	43210
    #define IP_ADDRESS 		"192.168.42.1"

Wireshark says the incoming client to drone UDP is ok, but ICMP d2c is not getting through:

    1335	2781.930156000	192.168.42.2	192.168.42.1	UDP	46	Source port: 55670  Destination port: 192
    1333	2781.416428000	192.168.42.1	192.168.42.2	ICMP	93	Destination unreachable (Port unreachable)

IO Buffers are as follows:

    {
        .ID = iobuffer_c2d_nack,
        .dataType = ARNETWORKAL_FRAME_TYPE_DATA,
        .sendingWaitTimeMs = 20, // 1
        .ackTimeoutMs = -1,
        .numberOfRetry = -1,
        .numberOfCell = 2,
        .dataCopyMaxSize = 128,
        .isOverwriting = 1,
    },
    {
        .ID = iobuffer_c2d_ack,
        .dataType = ARNETWORKAL_FRAME_TYPE_DATA_WITH_ACK,
        .sendingWaitTimeMs = 20,
        .ackTimeoutMs = 500,
        .numberOfRetry = 3,
        .numberOfCell = 20,
        .dataCopyMaxSize = 128,
        .isOverwriting = 0,
    },
    {
        .ID = iobuffer_c2d_emergency,
        .dataType = ARNETWORKAL_FRAME_TYPE_DATA_WITH_ACK,
        .sendingWaitTimeMs = 20, // 1
        .ackTimeoutMs = 100,
        .numberOfRetry = ARNETWORK_IOBUFFERPARAM_INFINITE_NUMBER,
        .numberOfCell = 1,
        .dataCopyMaxSize = 128,
        .isOverwriting = 0,
    }

And,

    {
        .ID = iobuffer_d2c_navdata,
        .dataType = ARNETWORKAL_FRAME_TYPE_DATA,
        .sendingWaitTimeMs = 20,
        .ackTimeoutMs = -1,
        .numberOfRetry = -1,
        .numberOfCell = 20,
        .dataCopyMaxSize = 128,
        .isOverwriting = 0,
    },
    {
        .ID = iobuffer_d2c_events,
        .dataType = ARNETWORKAL_FRAME_TYPE_DATA_WITH_ACK,
        .sendingWaitTimeMs = 20,
        .ackTimeoutMs = 500,
        .numberOfRetry = 3,
        .numberOfCell = 20,
        .dataCopyMaxSize = 128,
        .isOverwriting = 0,
    }

Thanks for any help.

**Comment by N-Bz** [comment]
Hi,

The connection to the bebop drone is done in three different steps.

1. Network discovery
 + The drone advertises itself on the network using an avahi daemon (you can search for the service using the bonjour protocol on OSX)
 + You get the drone ip address (typically 192.168.42.1) and the drone discovery tcp port (typically 44444) from this service
 + Note: This step can be omitted if you're sure of the values, but both information may change in later versions and doing this is recommanded.

2. ARDiscovery connection
 + Using the [ARDiscovery](https://github.com/ARDroneSDK3/libARDiscovery) library, and specifically the [ARDISCOVERY_Discovery](https://github.com/ARDroneSDK3/libARDiscovery/blob/master/Includes/libARDiscovery/ARDISCOVERY_Connection.h) class, you initiate a TCP connection with the drone.
 + During the ARDISCOVERY_Connection process, you will provide a JSON string (which will be sent to the drone) containing the following keys:
   - controller_type: The type (phone/tablet/computer ... or any value you want) of the controller
    - controller_name: The name of the controller
     - d2c_port: The UDP port you will listen to (i.e. given as "d2c" in the ARNETWORKAL_InitWifiNetwork function)
 + After this you will receive a JSON from the drone with the following keys:
   - c2d_port: The UDP port used by the drone to listen from ARNetwork (i.e. given as "c2d" in the ARNETWORKAL_InitWifiNetwork function)
    - Three arstream_xxx keys, used when you want to do video streaming
     - Maybe other keys that you can safely ignore at this point

3. ARNetwork connection
 + This is what you've done (and your code is correct, btw), but you can use the informations (ip / c2d_port / d2c_port) from the previous steps
 + If the c2d port sent by the drone is 0, it means that the drone refused the connection (probably because another device is already connected)

Basically, you're missing the ARDiscovery part. If the drone does not receive an ARDiscovery request, it won't send anything using ARNetwork.

Regards,

**Comment by simon-o-matic** [comment]
Thanks very much for that. I will add that and give it a try. Thought I didn't need that bit because I already knew the IP.

**Comment by weaversa** [comment]
@N-Bz If it's not too much trouble, would you be willing to give the same kind of rundown for connecting to the Rolling Spider?

**Comment by darryljackman** [comment]
Wow @N-Bz, that is more information that actually exist in the SDK!!! I can't believe how lacking this SDK is in documentation, explanations and examples.. UNACCEPTABLE!!

**Comment by CesarRN** [comment]
Hello all.

I have a similiar problem to simon-o-matic.


"

2015-03-11 13:33:56.330 RollingSpiderPiloting[2394:596324] -[ARDiscovery stop]:328
2015-03-11 13:34:00.912 RollingSpiderPiloting[2394:596438] start ...
[INF] DeviceController | 13:34:00:914 | -[DeviceController startNetwork]:411 - - Start ARNetwork
[INF] DeviceController | 13:34:00:914 | -[DeviceController ardiscoveryConnect]:370 - - ARDiscovery Connection
[DBG] DeviceController | 13:34:00:935 | ARDISCOVERY_Connection_ReceiveJsonCallback:1028 -     - ReceiveJson:{ "status": 0, "c2d_port": 54321, "arstream_fragment_size": 1000, "arstream_fragment_maximum_number": 128, "arstream_max_ack_interval": 0, "c2d_update_port": 51, "c2d_user_port": 21 } 
[ERR] ARNETWORK_Receiver | 13:34:02:722 | ARNETWORK_Receiver_ThreadRun:265 - data received, error: Insufficient free space in the buffer

"

And other, the famous:
"ARNETWORKAL_WifiNetwork_Receive:825 - connection lost (too long time without reception)"


I have the RollingSpiderPiloting project modified with Unix example and with the extract files from FreeFlight 3.

Any ideas ??

Thanks a lot!!
